File: code_examples/unsafe.cpp
#include <iostream>

int main() {
    int donuts[] = {1, 2, 3, 4, 5}; // Mmm... donuts!
    int* bart = donuts; // Bart's pointer to the donuts.

    std::cout << "Ay caramba!" << std::endl;

    for (int i = 0; i < 5; ++i) { // Bart's prank - doubling the donuts!
        // No bounds checking.
        *(bart + i) = *(bart + i) * 2; // Unsafe pointer arithmetic.
    }

    std::cout << "Uh oh! Bart's modified donuts: ";
    
    for (int i = 0; i < 5; ++i) {
        std::cout << donuts[i] << " ";
    }
    
    std::cout << "Woo-hoo!" << std::endl;

    return 0;
}

File: presentation_files/src/devconf_2024/16_9/code/unsafe.rs
fn main() {
    let mut donuts = [1, 2, 3, 4, 5];
    let bart = donuts.as_mut_ptr(); // Bart's pointer to the donuts.

    println!("Ay caramba!");

    unsafe {
        for i in 0..5 { // Bart's prank - doubling the donuts!
            // No bounds checking.
            *bart.add(i) *= 2; // Unsafe pointer arithmetic.
        }
    }

    print!("Uh oh! Bart's modified donuts: ");
    for i in 0..5 {
        print!("{} ", donuts[i]);
    }
    println!("Woo-hoo!");
}